cilium/CFP-37315-bgp-session-with-default-gateway.md

```diff
@@ -0,0 +1,165 @@
+
+config proposal:
+
+bgpInstances:
+  - name: "65001"
+    localASN: 65001
+    autoDiscovery: true
+    peers:
+    - name: "peer1 - 65000"
+      peerASN: 65000
+      afi: ipv4
+      peerConfigRef:
+        name: "cilium-peer"
+    - name: "peer2 - 65000"
+      peerASN: 65000
+      afi: ipv6
+      peerConfigRef:
+        name: "cilium-peer"
+
+In multi-homed environments (i.e., multiple default gateways leading to different ASNs), this proposal does not address which gateway belongs to which peer ASN.
+
+example multi-homed config:
+bgpInstances:
+  - name: "65001"
+    localASN: 65001
+    peers:
+    - name: "65000"
+      peerASN: 65000
+      peerConfigRef:
+        name: "cilium-peer"
+    - name: "65011"
+      peerASN: 65011
+      peerConfigRef:
+        name: "cilium-peer"
+
+If multiple default gateways exist, we cannot reliably match them to different peer ASNs. One workaround is to use same local ASN on both the ToRs for bgp sessions with the k8s node, and bgp cluster config will have one peer configured.
+
+proposed multi-homed config:
+bgpInstances:
+  - name: "65001"
+    localASN: 65001
+    peers:
+    - name: "65000"
+      peerASN: 65000
+      afi: ipv4
+      peerConfigRef:
+        name: "cilium-peer"
+
+proposed config in the device (arista):
+bgp listen range 10.2.0.0/24 peer-group SERVERS remote-as 65001
+router bgp 65100
+   router-id 10.1.1.1
+   timers bgp 1 3
+   neighbor SERVERS local-as 65000 no-prepend replace-as
+
+
+### Routes from Statedb
+
+cilium agent should rely on statedb to get the default gateway to create the bgp session. we can inject statedb into neighbor reconcillation struct and when we loop through neighbors, we can populate the peer ip by looking for default gateway of 0.0.0.0/0 or ::0/0 route in routes table of statedb.
+
+
+## Impacts / Key Questions
+
+### Impact 1: Multi-Homing Scenario
+
+In multi-homed environments, we will be breaking the design of one peer per bgp session because we will create two bgp sessions with one peer defined in the config. One of the workarounds could be to define two peers with same config and a unique name.
+
+### Key Question
+
+Is there a different way to accomplish this?
+
+
+### Impact 2: Multiple BGP Instances
+
+The proposal does not take into account multiple bgp instances so auto-discovery will work in that case.
+
+### Key Question
+
+What are use-cases of having multiple bgp instances? (bgp instances is a list in config)
+
+### Impact 3: config knob for auto-discovery
+
+config knob to enable auto-discovery will be crucial because that enables the users to understand the dependencies before making the changes. We can have documentation for auto-discover and provide the pre-requisites for enabling it.
+Without the knob, the user may use auto-discovery without understanding the requirements for it.
+
+### Key Question
+
+are we okay to add a knob for auto-discovery or should we enable auto-discovery by default if peerAddress is not specified?
```

Comment:
I'd prefer an explicit knob. Also, that implicit knob only allows single type of the discovery mechanism. We may want to add different kind of discovery mechanism in the future.

Reply:
perfect! 100% agreed!